mbuild/compound.py

```diff
@@ -759,48 +759,89 @@ def add(self, new_child, label=None, containment=True, replace=False,
-        self.children -= remove_from_here
-        yet_to_remove = objs_to_remove - remove_from_here
+        # Remove Port objects separately
+        ports_removed = set()
+        for obj in objs_to_remove:
+            if isinstance(obj, Port):
+                ports_removed.add(obj)
+                self._remove(obj)
+                obj.parent.children.remove(obj)
+                self._remove_references(obj)
+
+        objs_to_remove = objs_to_remove - ports_removed
+
+        # Get particles to remove
+        particles_to_remove = set([particle for obj in objs_to_remove
+                                            for particle in obj.particles()])
+
+        # Recursively get container compounds to remove
+        to_remove = list()
+
+        def _check_if_empty(child):
+            if child in to_remove:
+                return
+            if set(child.particles()).issubset(particles_to_remove):
+                if child.parent:
+                    to_remove.append(child)
+                    _check_if_empty(child.parent)
+                else:
+                    warn("This will remove all particles in "
+                            "compound {}".format(self))
+            return
 
-        for removed in remove_from_here:
-            for child in removed.children:
-                removed.remove(child)
+        for particle in particles_to_remove:
+            _check_if_empty(particle)
 
-        for removed_part in remove_from_here:
-            if removed_part.rigid_id is not None:
-                for ancestor in removed_part.ancestors():
-                    ancestor._check_if_contains_rigid_bodies = True
-            if self.root.bond_graph and self.root.bond_graph.has_node(
-                    removed_part):
-                for neighbor in self.root.bond_graph.neighbors(removed_part):
-                    self.root.remove_bond((removed_part, neighbor))
-                self.root.bond_graph.remove_node(removed_part)
+        # Fix rigid_ids and remove obj from bondgraph
+        for removed_part in to_remove:
+            self._remove(removed_part)
+
+        # Remove references to object
+        for removed_part in to_remove:
+            if removed_part.parent is not None:
+                removed_part.parent.children.remove(removed_part)
             self._remove_references(removed_part)
 
-        # Remove the part recursively from sub-compounds.
-        for child in self.children:
-            child.remove(yet_to_remove)
-            if child.contains_rigid:
-                self.root._reorder_rigid_ids()
+        # Remove ghost ports
+        all_ports_list = list(self.all_ports())
+        for port in all_ports_list:
+            if port.anchor not in [i for i in self.particles()]:
+                port.parent.children.remove(port)
+
+        # Check and reset rigid_id
+        if self.contains_rigid:
+            self.root._reorder_rigid_ids()
```

Comment:
This is going to fail because [`_reorder_rigid_ids()`](https://github.com/mosdef-hub/mbuild/blob/ed96251feaa119979f1916b5325e2ae14ad52898/mbuild/compound.py#L644-L664) only increments or decrements the rigid_ids by 1. So if we remove multiple rigid particles, then it will only account for only one of the removed particles. 

What I would recommend is to replace L830-831 with this:
```python
for _ in particles_to_remove:
    if self.contains_rigid:
        self.root._reorder_rigid_ids()
```

This ultimately is a limitation of the `_reorder_rigid_ids()` function that we should address in another PR.